File: backend/models/agent.py
"""AgentDefinition + AgentCreatePayload Pydantic models.

Field names align with the frontend `AgentDefinition` interface in src/services/api.ts
and Rust `AgentDefinition` in src-tauri/src/core/types.rs (snake_case, no rename_all).
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict


class AgentDefinition(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    name: str
    role: str
    system_prompt: str = ""
    skills: list[str] = []
    extra_skills: list[str] = []
    mounted_skills: list[str] = []
    allowed_tools: list[str] = []
    denied_tools: list[str] = []
    startup_strategy: str = ""
    model: str = ""
    max_turns: int = 0
    description: str | None = None
    metadata_: dict[str, Any] | None = None
    created_at: str = ""
    updated_at: str = ""


class AgentCreatePayload(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: str
    role: str
    system_prompt: str | None = None
    extra_skills: list[str] = []
    skills: list[str] = []
    mounted_skills: list[str] = []
    description: str | None = None

File: src-tauri/src/core/commands/agent.rs
//! Agent 命令 —— greenfield 重写
//! 参数名 camelCase（与前端 api.ts 对齐）

use crate::core::engine::registry;
use crate::core::store;
use crate::core::types::*;

#[tauri::command(rename_all = "camelCase")]
pub fn list_agents() -> Vec<AgentDefinition> {
    store::store().list_agents()
}

#[tauri::command(rename_all = "camelCase")]
pub fn get_agent(id: String) -> Option<AgentDefinition> {
    store::store().get_agent(&id)
}

#[tauri::command(rename_all = "camelCase")]
pub fn create_agent(payload: AgentCreatePayload) -> AgentDefinition {
    let now = store::now_iso();
    let agent = AgentDefinition {
        id: store::new_id("agent"),
        name: payload.name,
        role: payload.role,
        system_prompt: payload.system_prompt.unwrap_or_default(),
        skills: payload.skills,
        extra_skills: payload.extra_skills,
        allowed_tools: vec![],
        denied_tools: vec![],
        startup_strategy: "on_demand".into(),
        model: "glm-5.1".into(),
        max_turns: 50,
        description: payload.description,
        metadata: None,
        created_at: now.clone(),
        updated_at: now,
    };
    store::store().upsert_agent(agent)
}

#[tauri::command(rename_all = "camelCase")]
pub fn update_agent(id: String, payload: serde_json::Value) -> Option<AgentDefinition> {
    let agent = store::store().get_agent(&id)?;
    let mut merged = serde_json::to_value(&agent).ok()?;
    if let serde_json::Value::Object(map) = &mut merged {
        if let serde_json::Value::Object(patch) = payload {
            for (k, v) in patch {
                map.insert(k.clone(), v.clone());
            }
        }
        map.insert("updated_at".into(), serde_json::json!(store::now_iso()));
    }
    let updated: AgentDefinition = serde_json::from_value(merged).ok()?;
    Some(store::store().upsert_agent(updated))
}

#[tauri::command(rename_all = "camelCase")]
pub fn delete_agent(id: String) -> bool {
    store::store().delete_agent(&id)
}

/// 供 group 命令启动引擎用（内部）
#[allow(dead_code)]
pub fn ensure_engine(group_id: &str, agent_id: &str) {
    if let Some(agent) = store::store().get_agent(agent_id) {
        registry().add_engine(group_id, &agent);
    }
}
